### DISCUSSION TITLE
Can we set up a different dispatch predicate per data loader when using ScheduledDataLoaderRegistry?

### DISCUSSION THREAD

**Discussion by srinivasankavitha** [graphql-java/java-dataloader#125] [Q&A] [answered]
I'm trying to use the experimental ScheduledDataLoaderRegistry. We need to set up a different dispatch predicate per data loader. Is it possible to have multiple ScheduledDataLoaderRegistry instances, each with different dispatch predicates? 
There seems to be only one data loader registry that we can set up with the execution input. Is there another way to have a different dispatch predicate per data loader within the ScheduledDataLoaderRegistry?

```
ExecutionInput executionInput = newExecutionInput()
        .query(getQuery())
        .dataLoaderRegistry(registry)
        .build();
```

**Comment by bbakerman** [comment]
in graphql-java - no - there is only 1 dataloader registry used

What do you want multiple for - certain dataloaders get different parameters??

You could write a delegating DataLoaderRegistry that some how finds them from a group of them maybe??

**Reply by srinivasankavitha** [reply]
The use case is basically to allow a different dispatch predicate to be specified per data loader, since when and how often to dispatch can depend on the field being loaded. 

For context, we are using composed data loaders, and that requires a manual dispatch currently. To work around having to manually dispatch, we are looking to try out the ScheduledDispatchRegistry, but don't want the same dispatch predicate for all registered data loaders.

**Reply by bbakerman** [reply]
Maye something like

```
    class DelegatingDataLoaderRegistry extends DataLoaderRegistry {
        private final List<DataLoaderRegistry> registries;

        public DelegatingDataLoaderRegistry(List<DataLoaderRegistry> registries) {
            this.registries = registries;
        }

        @Override
        public List<DataLoader<?, ?>> getDataLoaders() {
            List<DataLoader<?, ?>> loaders = new ArrayList<>(super.getDataLoaders());
            registries.forEach( dlr -> loaders.addAll(dlr.getDataLoaders()));
            return loaders;
        }

        @Override
        public Map<String, DataLoader<?, ?>> getDataLoadersMap() {
            Map<String, DataLoader<?, ?>> loaders = new LinkedHashMap<>(super.getDataLoadersMap());
            registries.forEach( dlr -> loaders.putAll(dlr.getDataLoadersMap()));
            return loaders;
        }

        @Override
        public <K, V> DataLoader<K, V> getDataLoader(String key) {
            DataLoader<K, V> dataLoader = super.getDataLoader(key);
            if (dataLoader == null) {
                for (DataLoaderRegistry registry : registries) {
                    dataLoader = registry.getDataLoader(key);
                    if (dataLoader != null) {
                        return dataLoader;
                    }
                }
            }
            return null;
        }

        @Override
        public Set<String> getKeys() {
            return getDataLoadersMap().keySet();
        }

        @Override
        public void dispatchAll() {
            super.dispatchAll();
            for (DataLoaderRegistry registry : registries) {
                registry.dispatchAll();
            }
        }

        @Override
        public int dispatchAllWithCount() {
            int count = super.dispatchAllWithCount();
            for (DataLoaderRegistry registry : registries) {
                count += registry.dispatchAllWithCount();
            }
            return count;
        }

        @Override
        public int dispatchDepth() {
            int count = super.dispatchDepth();
            for (DataLoaderRegistry registry : registries) {
                count += registry.dispatchDepth();
            }
            return count;
        }

    }
```

where the main registry is really a list of others ??

This is a rough bit of code - eg should you be alllowed to "register new DL" into this combining one??  Or should computeIfAnsent work or not?

I am not sure - I just whipped this up as an idea

**Reply by srinivasankavitha** [reply]
Thanks so much. I think this makes the most sense for our use case. Will try this out. Do you think this would be something worth supporting in the future as part of the java-dataloader? Or is it too specific a use case?

**Reply by bbakerman** [reply]
Nah perhaps it would be a good idea

**Reply by LeiWangHoward** [reply]
Any specific reason we don't want to support per dataloader dispatch predicate configuration? 
- The current design of `DispatchPredicate` is already testing the criteria/predicate per-dataLoader: https://github.com/graphql-java/java-dataloader/blob/master/src/main/java/org/dataloader/registries/DispatchPredicate.java#L21
- Use case for the per-dataLoader predicate configuration: different batch function may have different limit based on upstream API (e.g. `myBatchGetUsers` might have maximum user limit 5, thus I want to configure queue depth predicate to 5 for `myBatchGetUsers` dataLoader, while have queue depth predicate set to 10 for other dataLoaders.
- Netflix DGS supports this feature: https://netflix.github.io/dgs/data-loaders/#scheduled-data-loaders-with-dispatch-predicates Potentially we can have `DispatchPredicate` as a parameter of BatchLoader

Any plan to support this feature, or it's okay to submit community pull request?

**Reply by bbakerman** [reply]
This is now supported in 3.2.1

**Comment by bbakerman** [comment, marked as answer]
See https://github.com/graphql-java/java-dataloader/pull/130/files

**Reply by srinivasankavitha** [reply]
This is great, thank you!
